src/Client/HedgedConnectionsFactory.cpp

```diff
@@ -0,0 +1,447 @@
+    }
+}
+
+int HedgedConnectionsFactory::getReadyFileDescriptor(bool blocking)
+{
+    for (auto & [fd, replica_index] : fd_to_replica_index)
+        if (connection_establishers[replica_index].result.entry->hasReadPendingData())
+            return connection_establishers[replica_index].socket_fd;
+
+    epoll_event event;
+    event.data.fd = -1;
+    epoll.getManyReady(1, &event, blocking);
+    return event.data.fd;
+}
+
+HedgedConnectionsFactory::State HedgedConnectionsFactory::processReplicaEvent(int replica_index, Connection *& connection_out)
+{
+    removeTimeoutsFromReplica(replica_index);
+    connection_establishers[replica_index].run();
+    State state = processConnectionEstablisherStage(replica_index, true);
+    if (state == State::NOT_READY)
+        addTimeouts(replica_index);
+    if (state == State::READY)
+        connection_out = &*connection_establishers[replica_index].result.entry;
+    return state;
+}
+
+bool HedgedConnectionsFactory::processTimeoutEvent(int replica_index, ConnectionTimeoutDescriptorPtr timeout_descriptor)
+{
+    epoll.remove(timeout_descriptor->timer.getDescriptor());
+    replicas_timeouts[replica_index].erase(timeout_descriptor->timer.getDescriptor());
+    timeout_fd_to_replica_index[timeout_descriptor->timer.getDescriptor()];
+
+    if (timeout_descriptor->type == ConnectionTimeoutType::RECEIVE_TIMEOUT)
+    {
+        removeTimeoutsFromReplica(replica_index);
+        int fd = connection_establishers[replica_index].socket_fd;
+        epoll.remove(fd);
+        fd_to_replica_index.erase(fd);
+
+        ConnectionEstablisher & connection_establisher = connection_establishers[replica_index];
+        connection_establisher.fail_message = "Receive timeout expired (" + connection_establisher.result.entry->getDescription() + ")";
+        connection_establisher.resetResult();
+        connection_establisher.stage = ConnectionEstablisher::Stage::FAILED;
+        processFailedConnection(replica_index);
+        return true;
+    }
+
+    /// Return true if we can try to start one more connection.
+    return entries_count + indexes_in_process.size() + failed_pools_count < shuffled_pools.size();
+}
+
+HedgedConnectionsFactory::State HedgedConnectionsFactory::setBestUsableReplica(Connection *& connection_out)
+{
+    std::vector<int> indexes(connection_establishers.size());
+    for (size_t i = 0; i != indexes.size(); ++i)
+        indexes[i] = i;
+
+    /// Remove unusable, failed replicas and replicas that are ready or in process.
+    indexes.erase(
+        std::remove_if(
+            indexes.begin(),
+            indexes.end(),
+            [&](int i)
+            {
+                return connection_establishers[i].result.entry.isNull() || !connection_establishers[i].result.is_usable ||
+                    indexes_in_process.find(i) != indexes_in_process.end() || ready_indexes.find(i) != ready_indexes.end();
+            }),
+        indexes.end());
```

Comment:
It may be not very important, but why we first insert into vector, then erase from it?
Why not, just, push_back in loop?

Reply:
Also, maybe create a struct for current replica status? And replace `indexes_in_process` and `ready_indexes` to `std::vector<ReplicaStatus>` and a couple of counters?

Reply:
Probably, `ConnectionEstablisher` will be a part of this struct.